File: src/main/java/org/cobbzilla/util/io/CompositeBufferedFilesystemWatcher.java
package org.cobbzilla.util.io;

import java.io.File;
import java.nio.file.Path;
import java.nio.file.WatchEvent;
import java.util.Collection;
import java.util.List;

public abstract class CompositeBufferedFilesystemWatcher extends CompositeFilesystemWatcher<BufferedFilesystemWatcher> {

    private final long timeout;
    private final int maxEvents;

    public CompositeBufferedFilesystemWatcher (long timeout, int maxEvents, File[] paths) {
        this(timeout, maxEvents);
        addAll(paths);
    }

    public CompositeBufferedFilesystemWatcher (long timeout, int maxEvents, String[] paths) {
        this(timeout, maxEvents);
        addAll(paths);
    }

    public CompositeBufferedFilesystemWatcher (long timeout, int maxEvents, Path[] paths) {
        this(timeout, maxEvents);
        addAll(paths);
    }

    public CompositeBufferedFilesystemWatcher (long timeout, int maxEvents, Collection things) {
        this(timeout, maxEvents);
        addAll(things);
    }

    @java.beans.ConstructorProperties({"timeout", "maxEvents"})
    public CompositeBufferedFilesystemWatcher(long timeout, int maxEvents) {
        this.timeout = timeout;
        this.maxEvents = maxEvents;
    }

    public abstract void fire(List<WatchEvent<?>> events);
    private void _fire(List<WatchEvent<?>> events) { fire(events); }

    @Override protected BufferedFilesystemWatcher newWatcher(Path path) {
        return new BufferedFilesystemWatcher(path, timeout, maxEvents) {
            @Override protected void fire(List<WatchEvent<?>> events) {
                _fire(events);
            }
        };
    }

    public java.lang.String toString() {
        return "CompositeBufferedFilesystemWatcher(super=" + super.toString() + ", timeout=" + this.timeout + ", maxEvents=" + this.maxEvents + ")";
    }

    public long getTimeout() {
        return this.timeout;
    }

    public int getMaxEvents() {
        return this.maxEvents;
    }
}

File: src/main/java/org/cobbzilla/util/io/CompositeBufferedFilesystemWatcher.kt
package org.cobbzilla.util.io

import java.io.File
import java.nio.file.Path
import java.nio.file.WatchEvent

abstract class CompositeBufferedFilesystemWatcher @java.beans.ConstructorProperties("timeout", "maxEvents")
constructor(val timeout: Long, val maxEvents: Int) : CompositeFilesystemWatcher<BufferedFilesystemWatcher>() {

    constructor(timeout: Long, maxEvents: Int, paths: Array<File>) : this(timeout, maxEvents) {
        addAll(paths)
    }

    constructor(timeout: Long, maxEvents: Int, paths: Array<String>) : this(timeout, maxEvents) {
        addAll(paths)
    }

    constructor(timeout: Long, maxEvents: Int, paths: Array<Path>) : this(timeout, maxEvents) {
        addAll(paths)
    }

    constructor(timeout: Long, maxEvents: Int, things: Collection<*>) : this(timeout, maxEvents) {
        addAll(things)
    }

    abstract fun fire(events: List<WatchEvent<*>>)
    private fun _fire(events: List<WatchEvent<*>>) {
        fire(events)
    }

    override fun newWatcher(path: Path): BufferedFilesystemWatcher {
        return object : BufferedFilesystemWatcher(path, timeout, maxEvents) {
            override fun fire(events: List<WatchEvent<*>>) {
                _fire(events)
            }
        }
    }

    override fun toString(): java.lang.String {
        return "CompositeBufferedFilesystemWatcher(super=" + super.toString() + ", timeout=" + this.timeout + ", maxEvents=" + this.maxEvents + ")"
    }
}
